text/0000-add-filename-info-to-io-error.md

```diff
@@ -0,0 +1,148 @@
+
+```txt
+error: No such file or directory
+ --> src/main.rs:4:46
+  |
+4 |     let hello_file = File::open("hello.txt")?;
+  |                                 ^^^^^^^^^^ `hello.txt` doesn't exist
+  |
+  = help: create a file named `hello.txt` in your current directory
+
+error: aborting due to previous error
+```
+
+This error explains where the error is emitted and the cause of it. Also, this error adds a help message to create a `hello.txt` file in the current directory which would make this program compile successfully. Let's look at another case:
+
+```rust
+use std::fs::File;
+
+fn main() -> std::io::Result<()> {
+    let _ = File::create("/etc/protocols")?;
+
+    Ok(())
+}
+```
+
+This program returns the following error:
+
+```txt
+Error: Os { code: 13, kind: PermissionDenied, message: "Permission denied" }
+```
+
+This program is not helpful at all, not displaying file information or how to fix this error. With this change, here is the projected error message:
+
+```
+error: Permission denied
+ --> src/main.rs:4:36
+  |
+4 |     let _ = File::create("/etc/protocols")?;
+  |                          ^^^^^^^^^^ permission denied
+  |
+  = help: give write access to this program
+  = help: try running this command: `sudo chown -R $USER /etc/protocols`
+
+error: aborting due to previous error
+```
+
+In this case, this change would add the file to the `Os` error message. Many crates and Rust applications write their own error messages to improve the experience while working with `std::io`. This is why we should incorporate filename information to `std::io:Error` to reduce the amount of boilerplate crates and Rust applications need to write to ensure a better `std::io` development process.
+
+# Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+1\.
+
+- Add field `file: Option<PathBuf>` to struct `std::io::Error`
+  - Whenever a file doesn't exist or another error occurs when writing/reading/accessing a file, add the filename to the `file` field
+
+2\.
+
+- Test impact of performance by adding benchmarks and looking at `std`'s and `std::io`'s general difference in performance before and after this implementation
+
+3\.
+
+- Create a diagnostic such as above
+  - Make the lint as helpful as possible, adding `help` messages and pointing to where in the code the error is being caused
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- `Option<PathBuf>` would make `std::io:Error` heavier than before [(originally pointed out by @withoutboats)](https://internals.rust-lang.org/t/add-filename-information-to-io-error/5120/5)
+  - In some cases, this would also make `std::io:Error` inefficient as operations which expect a large number of failures would have to allocate to the error itself and the `PathBuf` in the `file` field [(originally pointed out by @alexcrichton)](https://internals.rust-lang.org/t/add-filename-information-to-io-error/5120/7)
+
+# Rationale and alternatives
```

Comment:
An alternative is to attach the filename and the operation done to the error by adding context to it, as is widely applied by developers using crates such as `anyhow`.